### PR TITLE
Add support for MSC4717: service members 🫡

### PR DESCRIPTION
**Pull request by poljar**
Depends on https://github.com/matrix-org/matrix-rust-sdk/pull/4228, only the second commit here is relevant.

This adds support for service members to be excluded from the room display name calculation logic as described in [MSC4717](https://github.com/matrix-org/matrix-spec-proposals/blob/tulir/service-members/proposals/4171-service-members.md).

### CODE REVIEW

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (poljar) [reply crates/matrix-sdk-base/src/rooms/normal.rs:239]: This is just a standard partially applied function, nothing really fancy going on.

But we're not interested in the members in both cases, in the case where we do have a summary we're only fetching individual members for which we didn't receive the name of the hero as part of the summary:

https://github.com/matrix-org/matrix-rust-sdk/blob/0ecbbc2e6e1bce101dc1b59efffd6a0276552362/crates/matrix-sdk-base/src/rooms/normal.rs#L706-L722

Comment (bnjbvr) [reply crates/matrix-sdk-base/src/rooms/normal.rs:239]: Ohh you're right, members are not loaded in this case.

(It might also mean that if you have a room with 10 members, and the summary retrieved from the server contains only 6 but it includes the service member, then the final room name will include 5 heroes, while it could include another non-service-member? Maybe a feedback for the MSC, unless the server knows about service members.)

Could we have a freestanding function iterating over user_ids and returning an `impl Iterator<Item = &UserId>` that contains the filtered ids, maybe?

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (poljar) [reply crates/matrix-sdk-base/src/rooms/normal.rs]: Alright, changed it to a `-` operation.

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (poljar) [reply crates/matrix-sdk-base/src/rooms/normal.rs]: Moved this to another PR: https://github.com/matrix-org/matrix-rust-sdk/pull/4353

**Comment by bnjbvr** [comment]
(Removed review request until the dependent PR is merged first.)

**Comment by poljar** [comment]
> (Removed review request until the dependent PR is merged first.)

Why?

**Comment by bnjbvr** [comment]
Because OP reads:

> Depends on https://github.com/matrix-org/matrix-rust-sdk/pull/4228

**Comment by poljar** [comment]
> Because OP reads:
> 
> > Depends on #4228

Ok but you reviewed the PR even though OP read that same message. Then I addressed all of your review comments except one.

Now you're telling me that you won't review it because OP reads this message. I'm asking why you changed your mind and why do you think that #4228 needs to be merged before this one is reviewed?

They are conceptually unrelated, they are tied together because Ruma had breaking changes.

On the topic of #4228, seems that the author also addressed your concerns but is as well waiting on your review.

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (bnjbvr) [reply crates/matrix-sdk-base/src/rooms/normal.rs:239]: > Why?

A detour for programmers, which is cognitive overhead; more complicated type checking for the compiler; I do prefer simplicity in the form of having a single line repeated in two places, rather than a new function returning a common filter — sometimes, the good intent of sharing code ends up confusing the reader a bit too much. But it's alright, as you seem to have a strong opinion on this, and this ain't the hill I want to die on :+1:

Comment (poljar) [reply crates/matrix-sdk-base/src/rooms/normal.rs:239]: >But it's alright, as you seem to have a strong opinion on this, and this ain't the hill I want to die on 👍

I mean you didn't suggest any alternative, the first thing you suggested didn't work, the second suggestions I'm fishing for meaning but I don't really know what you meant.

> I do prefer simplicity in the form of having a single line repeated in two places

Ok now you do suggest something, and I have to disagree here. I almost missed that the filtering needs to happen in two places, if we do this and somebody else has to modify the filter they'll very likely miss this fact.

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:704]: This may or may not create a the inner closure multiple times (once for every call to the outer `heroes_filter`). We can construct it once explicitly instead, with:

```rust
let parent_filter = heroes_filter(own_user_id, &member_hints);
let heroes_filter = |hero: &&RoomHero| parent_filter(&hero.user_id);
```

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:738]: Same here.

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (bnjbvr) [inline crates/matrix-sdk-base/src/rooms/normal.rs:798]: (light suggestion) A bit more concise, and avoids the chaining of `and_then` twice + `map`:

```rust
            .and_then(|event| as_variant!(event, SyncOrStrippedState::Sync(SyncStateEvent::Original(e)) => e.content))
```

**Review by bnjbvr — APPROVED** [review]
Thanks! Only reviewed the second commit, assuming we'll wait for the dependent PR to be first reviewed and merged before landing this.

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (poljar) [reply crates/matrix-sdk-base/src/rooms/normal.rs]: 3c12242e22fce2996410f2f4d75dd732c9fa5bd8

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (poljar) [reply crates/matrix-sdk-base/src/rooms/normal.rs]: a8827386ea6bc9462f2f8faffad49c7740096ac8

File: crates/matrix-sdk-base/src/rooms/normal.rs
Comment (poljar) [reply crates/matrix-sdk-base/src/rooms/normal.rs]: a8827386ea6bc9462f2f8faffad49c7740096ac8

**Pull request merged**
